cmd/opampsupervisor/supervisor/packages.go

```diff
@@ -0,0 +1,474 @@
+		restoreErr := restoreBackup(agentBackupPath, p.agentPath)
+		return errors.Join(fmt.Errorf("write package to file: %w", err), restoreErr)
+	default:
+		restoreErr := restoreBackup(agentBackupPath, p.agentPath)
+		return errors.Join(fmt.Errorf("agent package met or exceeded %d bytes", maxAgentBytes), restoreErr)
+	}
+
+	return nil
+}
+
+func (p *packageManager) DeletePackage(packageName string) error {
+	if packageName != agentPackageKey {
+		// We only take the agent package, so the package already doesn't exist.
+		return nil
+	}
+
+	// We will never delete the agent package.
+	return errors.New("cannot delete top-level package")
+}
+
+func (p *packageManager) LastReportedStatuses() (*protobufs.PackageStatuses, error) {
+	lastStatusBytes, err := os.ReadFile(p.lastPackageStatusPath())
+	switch {
+	case errors.Is(err, os.ErrNotExist):
+		// No package statuses exists
+		return nil, nil
+	case err != nil:
+		return nil, fmt.Errorf("read last package statuses: %w", err)
+	}
+
+	var ret protobufs.PackageStatuses
+	err = proto.Unmarshal(lastStatusBytes, &ret)
+	if err != nil {
+		return nil, fmt.Errorf("unmarshal last package statuses: %w", err)
+	}
+
+	return &ret, nil
+}
+
+func (p packageManager) SetLastReportedStatuses(statuses *protobufs.PackageStatuses) error {
+	lastStatusBytes, err := proto.Marshal(statuses)
+	if err != nil {
+		return fmt.Errorf("marshal statuses: %w", err)
+	}
+
+	err = os.WriteFile(p.lastPackageStatusPath(), lastStatusBytes, 0600)
+	if err != nil {
+		return fmt.Errorf("write package statues: %w", err)
+	}
+
+	return nil
+}
+
+func (p *packageManager) lastPackageStatusPath() string {
+	return filepath.Join(p.storageDir, lastPackageStatusFileName)
+}
+
+func (p *packageManager) packagesStatusPath() string {
+	return filepath.Join(p.storageDir, packagesStateFileName)
+}
+
+func loadPackageState(path string) (*packageState, error) {
+	by, err := os.ReadFile(path)
+	if err != nil {
+		return nil, fmt.Errorf("read file: %w", err)
+	}
+
+	var pkgState packageState
+	if err := yaml.Unmarshal(by, &pkgState); err != nil {
+		return nil, fmt.Errorf("unmarshal yaml: %w", err)
+	}
+
+	return &pkgState, nil
+}
+
+func savePackageState(path string, ps *packageState) error {
+	by, err := yaml.Marshal(ps)
+	if err != nil {
+		return fmt.Errorf("marshal yaml: %w", err)
+	}
+
+	f, err := os.OpenFile(path, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0600)
+	if err != nil {
+		return fmt.Errorf("open file: %w", err)
+	}
+	defer f.Close()
+
+	if _, err := f.Write(by); err != nil {
+		return fmt.Errorf("write package state: %w", err)
+	}
+
+	return nil
+}
+
+func verifyPackageIntegrity(packageBytes, expectedHash []byte) error {
+	actualHash := sha256.Sum256(packageBytes)
+	if !bytes.Equal(actualHash[:], expectedHash) {
+		return errors.New("invalid hash for package")
+	}
+
+	return nil
+}
+
+func parsePackageSignature(signature []byte) (b64Cert, b64Signature []byte, err error) {
+	splitSignature := bytes.SplitN(signature, []byte(" "), 2)
```

Comment:
Is this signature format documented somewhere? This imposes a requirement on the OpAMP server to generate signatures in a specific way, right?